prune_shard_list.c

```diff
@@ -239,6 +249,51 @@ MakeOpExpression(Var *variable, int16 strategyNumber)
 
 
 /*
+ * LookupOperatorByType is a wrapper around GetOperatorByType that uses a cache
+ * to avoid multiple lookups of operators by their types.
+ */
+static Oid
+LookupOperatorByType(Oid typeId, Oid accessMethodId, int16 strategyNumber)
+{
+	OperatorTypeCacheEntry *matchingCacheEntry = NULL;
+	ListCell *cacheEntryCell = NULL;
+
+	/* search the cache */
+	foreach(cacheEntryCell, OperatorTypeCache)
+	{
+		OperatorTypeCacheEntry *cacheEntry = lfirst(cacheEntryCell);
+
+		if (cacheEntry->typeId == typeId && cacheEntry->accessMethodId == accessMethodId
+			&& cacheEntry->strategyNumber == strategyNumber)
+		{
+			matchingCacheEntry = cacheEntry;
+			break;
+		}
+	}
+
+	/* if not found in the cache, call GetOperatorByType and put the result in cache */
+	if (matchingCacheEntry == NULL)
+	{
+		Oid operatorId = GetOperatorByType(typeId, accessMethodId, strategyNumber);
+		MemoryContext oldContext = MemoryContextSwitchTo(CacheMemoryContext);
```

Comment:
I'm not sure when the inputs might cause this, but `GetOperatorByType` will return `InvalidOid` if the type isn't found. In that case, we don't want to cache anything, so you should check for that case and avoid caching a value. While it's _unlikely_ that we could get `InvalidOid` during one query and then find a valid operator the next, it's probably not 100% impossible, so it'd be good to do.

See [analogous code][1] in distribution_metadata.c for a pattern to follow.

[1]: https://github.com/citusdata/pg_shard/blob/9fa64edd98858ef1a3eaaca460deab7157dc122e/distribution_metadata.c#L90-L115